### PR TITLE
Create Shadow Property for FK

### LINKED ISSUE #3576: Shadow property is not created by .ForeignKey(Type, string)
... and is not taken into account if defined late in the Model.

Repro: this code should not throw an exception, but actually complains that `DocumentId` is not found or defined as a shadow property

```c#
    class Program
    {
        static void Main(string[] args)
        {
            var builder = new DbContextOptionsBuilder();
            var conventionBuilder = new CoreConventionSetBuilder();
            var modelBuilder = new ModelBuilder(
                conventionBuilder.CreateConventionSet()
                );

            modelBuilder.Entity<Document>(x =>
            {
                // I want the DocumentId FK on the PersonByName table
                x.HasOne(typeof(PersonByName), "PersonByName")
                    .WithOne()
                    .ForeignKey(typeof(PersonByName), "DocumentId");
            });

            modelBuilder.Entity<PersonByName>(x =>
            {
                // Defines shadow property DocumentId explicitely
                x.Property(typeof(int), "DocumentId");
            });

            modelBuilder.Validate();

            builder.UseModel(modelBuilder.Model);
            var dbFileName = Path.GetTempFileName();
            builder.UseSqlite(@"Data Source=" + dbFileName + ";Cache=Shared");

            // Ensure the database is created
            var dbContext = new DbContext(builder.Options);
            dbContext.Database.EnsureCreated();

            try {
                var p1 = new PersonByName { Name = "Bill" };
                var d1 = new Document { PersonByName = p1 };
                dbContext.Set<PersonByName>().Add(p1);
                dbContext.Set<Document>().Add(d1);
                dbContext.SaveChanges();

                IQueryable<Document> documents = dbContext.Set<Document>();
                Console.WriteLine(1 == documents.Count());

                Console.WriteLine(null != documents
                    .Where(x => x.PersonByName.Name == "Bill").FirstOrDefault());

               
            }
            finally
            {
                dbContext.Dispose();
                File.Delete(dbFileName);
            }
        }
    }

    public class Document
    {
        public int Id { get; set; }
        public PersonByName PersonByName { get; set; }
    }

    public class PersonByName
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }
```

### PR DESCRIPTION
**Pull request by smitpatel**
fixes #3576

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 1]
```diff
@@ -861,9 +861,25 @@ private static IndexBuildersSnapshot DetachIndexes(IEnumerable<Index> indexesToD
...
+                var newRelationshipBuilder = relationshipBuilder.HasForeignKey(propertyNames, configurationSource);
+                if (newRelationshipBuilder == null)
+                {
+                    RemoveForeignKey(relationshipBuilder.Metadata, configurationSource);
+                    return null;
+                }
+                return newRelationshipBuilder;
+            }
+
+            return HasForeignKeyInternal(principalType, GetOrCreateProperties(propertyNames, configurationSource), configurationSource);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs]: You could unify this new code with the code in `HasForeignKey` by moving it to `GetOrCreateProperties` and adding a parameter of `IEnumerable<Type>` to it containing the principal properties CLR types

**Comment by AndriySvyryd** [comment]
:shipit:

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs (hunk 1 above)
Comment (smitpatel) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs]: Should we create properties when the principal key is shadow?

Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs]: I think we decided we wouldn't. You can pass a null if the corresponding principal properties is shadow.

**Comment by smitpatel** [comment]
@AndriySvyryd - Can use another look at the PR.

File: src/EntityFramework.Core/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 2]
```diff
@@ -902,11 +902,23 @@ private bool CanInvert(ConfigurationSource configurationSource)
...
+            var dependentEntityTypeBuilder = ModelBuilder.Entity(Metadata.DeclaringEntityType.Name, configurationSource);
+            var principalProperties = Metadata.PrincipalKey.Properties;
+
+            var properties = dependentEntityTypeBuilder
+                .GetOrCreateProperties(
+                    propertyNames,
+                    configurationSource,
+                    principalProperties.Count != propertyNames.Count
+                    ? null
+                    : principalProperties.Select(p => p.IsShadowProperty ? null : p.ClrType));
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalRelationshipBuilder.cs]: It looks like you can further reduce duplication by just passing through the principal properties and moving the filtering logic to `GetOrCreateProperties`

**Comment by AndriySvyryd** [comment]
Looks neat
:shipit:

**Pull request merged**